Add table of available variables to README.

keithrbennett opened the pull request:
This puts the available variables in the readme so that the reader can see what variables are available without consulting the source code.

keithrbennett committed:
Add table of available variables to README.

keithrbennett committed:
Remove extra blank line.

keithrbennett committed:
Make markdown source more consistent.

olleolleolle commented:
I find this helpful and kind to the reader who navigates this repo.

The person navigating a rendered API will already see the names in some index or generated list (I guess).

fwolfst commented:
Additionally, a link to the implementation (https://github.com/ruby/English/blob/master/lib/English.rb) could be added to the README.

keithrbennett committed:
Merge branch 'master' into add-vars-to-readme

keithrbennett committed:
In readme, sort alias/built-in pairs alphabetically by alias.

keithrbennett committed:
Add link to english.rb as per @fwolfst's suggestion.

keithrbennett committed:
Add exe/alias_markdown_table_generator script.

keithrbennett commented:
I've made the following changes to this pull request branch:

* Add link to English.rb as per @fwolfst 's suggestion
* In readme, sort alias/built-in pairs alphabetically by alias
* Slightly improve presentation of the information
* Add script to generate the markdown table for the readme.

To the owner(s) of this repo:

* This PR is quite simple, does not affect runtime behavior, and would be helpful to users. Would you please take a look at it and specify any required changes, accept/reject it, etc.?
*  I realize the script adds code that would need to be considered/maintained going forward, and can remove it if you want.

keithrbennett committed:
Fix centering of built-in column text in rendered markdown; minor refactoring.

keithrbennett committed:
Rename entry point method.

nobu commented on exe/alias_markdown_table_generator:
```diff
@@ -0,0 +1,96 @@
+
+class AliasMarkdownTableGenerator
+  class << self
+    ALIAS_SOURCE_FILESPEC = File.absolute_path(File.join(File.basename(__FILE__), '../lib/English.rb'))
```
Did you intend `File.dirname(__FILE__)`?

```suggestion
    ALIAS_SOURCE_FILESPEC = File.join(__dir__, 'lib/English.rb')
```

keithrbennett replied:
You're right, `File.join(File.basename(__FILE__), '../lib/English.rb')` evaluates to `alias_markdown_table_generator/../lib/English.rb`, which makes no sense. Remarkably though, when that is wrapped in a `File.absolute_path()`, it returns the correct directory, which is why I did not catch that error.

I agree with your suggestion to replace with `__dir__`.

D-system commented:
This information is very helpful. 
Thankfully, as of today, it's the only PR therefore it is easy to find.
Even with only the hard-coded text in readme, it will be helpful.

keithrbennett committed:
Merge branch 'master' into add-vars-to-readme

keithrbennett committed:
Regenerate readme table to remove $IGNORECASE removed variable.

keithrbennett committed:
Add 'exe' gemspec section back due to introduction of alias_markdown_table_generator script.

keithrbennett commented:
I have just merged the changes made in the master branch since I posted this PR fifteen months ago. I also reran the table generator script to update the readme. (`$IGNORECASE` was removed from the code and therefore needed to be removed from the documentation.)

There is a table mapping English names to their terse built-in counterparts but not vice versa (i.e. `$ARGV` -> `$*` but not `$*` -> `$ARGV`). This would be especially helpful in cases containing multiple aliases. I'm willing to add the reverse table mapping if there's interest in moving forward with this PR.

It might be argued that adding this table risks it becoming out of date with respect to the code, but that is already the case with the comment in the .rb file. In addition, placing this information in the readme instead is more customary and more likely to be helpful to someone not already familiar with this code. To date, this change has received support from two community members (@olleolleolle and @D-system) with no opposing viewpoints.

If you still believe this change is ill-advised, then it would be helpful to me (and possibly others) to understand why, and have the opportunity to address any issues, if possible. I plan to be at Ruby Kaigi next week to discuss this with anyone interested.

D-system commented:
I'll also be at RubyKaigi 2024 ... I'm already in Okinawa, I prefer to arrive early for the best date of the year ✌️
Thanks @keithrbennett for the PR update

keithrbennett committed:
Refactor script to separate file parsing from text formatting.

keithrbennett committed:
Add 'builtin to alias' table to supplement existing 'alias to builtin' table.

keithrbennett commented:
I have refactored and extended the markdown table generator script to create tables in both directions, that is (Alias -> Builtin) and (Builtin -> Alias). The README.md is updated in this branch (see https://github.com/keithrbennett/English/blob/add-vars-to-readme/README.md) to include that table output. This is copied from that readme:

## Builtin to Alias Table

| Builtin | Alias                    |
| :-------: | ----------------------   |
|   $!    | $ERROR_INFO              |
|   $$    | $PID                     |
|   $$    | $PROCESS_ID              |
|   $&    | $MATCH                   |
|   $'    | $POSTMATCH               |
|   $*    | $ARGV                    |
|   $+    | $LAST_PAREN_MATCH        |
|   $,    | $OFS                     |
|   $,    | $OUTPUT_FIELD_SEPARATOR  |
|   $.    | $INPUT_LINE_NUMBER       |
|   $.    | $NR                      |
|   $/    | $INPUT_RECORD_SEPARATOR  |
|   $/    | $RS                      |
|   $;    | $FIELD_SEPARATOR         |
|   $;    | $FS                      |
|   $<    | $DEFAULT_INPUT           |
|   $>    | $DEFAULT_OUTPUT          |
|   $?    | $CHILD_STATUS            |
|   $@    | $ERROR_POSITION          |
|   $\    | $ORS                     |
|   $\    | $OUTPUT_RECORD_SEPARATOR |
|   $_    | $LAST_READ_LINE          |
|   $`    | $PREMATCH                |
|   $~    | $LAST_MATCH_INFO         |

## Alias to Builtin Table

| Alias                    | Builtin |
| ------------------------ | :-----: |
| $ARGV                    |   $*    |
| $CHILD_STATUS            |   $?    |
| $DEFAULT_INPUT           |   $<    |
| $DEFAULT_OUTPUT          |   $>    |
| $ERROR_INFO              |   $!    |
| $ERROR_POSITION          |   $@    |
| $FIELD_SEPARATOR         |   $;    |
| $FS                      |   $;    |
| $INPUT_LINE_NUMBER       |   $.    |
| $INPUT_RECORD_SEPARATOR  |   $/    |
| $LAST_MATCH_INFO         |   $~    |
| $LAST_PAREN_MATCH        |   $+    |
| $LAST_READ_LINE          |   $_    |
| $MATCH                   |   $&    |
| $NR                      |   $.    |
| $OFS                     |   $,    |
| $ORS                     |   $\    |
| $OUTPUT_FIELD_SEPARATOR  |   $,    |
| $OUTPUT_RECORD_SEPARATOR |   $\    |
| $PID                     |   $$    |
| $POSTMATCH               |   $'    |
| $PREMATCH                |   $`    |
| $PROCESS_ID              |   $$    |
| $RS                      |   $/    |

keithrbennett committed:
Remove comment no longer applicable.

keithrbennett committed:
Add table headings to markdown table generator.

keithrbennett committed:
Change string output to HEREDOC, update sample output.

hsbt commented on English.gemspec:
```diff
@@ -19,4 +19,6 @@ Gem::Specification.new do |spec|
     `git ls-files -z 2>#{IO::NULL}`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
   end
   spec.require_paths = ["lib"]
+  spec.bindir        = "exe"
```
Can you remove this and next line? These configuration is for distribute executables for gem users. `exe/alias_markdown_table_generator` seems for developer utility, not `English` users.

keithrbennett replied:
Yes, thank you, that's a great point. I have removed those lines.

nobu replied:
I think this file should be under `bin/`, not `exe/`.

keithrbennett committed:
Remove exe lines from gemspec as per @hsbt pointing out that gem users do not need the executable script.

keithrbennett committed:
Change output heading from ## to ### to indicate that these tables are a part of the Usage section of the readme.
Remove lengthy sample output from script comment, and instead refer to the output in comment at top of file.

keithrbennett committed:
Change parse method param to be array of source code text lines instead of a filename, to make the method more general and easier to test/experiment/etc.

hsbt commented:
@keithrbennett Thanks for your work! I merged this because it is enough for users. If you want to add extra work, Please open another PRs.

keithrbennett commented:
Thank you, too, @hsbt . Feel free to delete the `add-vars-to-readme` branch if you wish. (I was going to do it but then thought that you might want to make that decision yourself.)

I don't plan to add anything to this change.

nobu commented on exe/alias_markdown_table_generator:
```diff
@@ -0,0 +1,86 @@
+  puts <<~OUTPUT
+    ### Builtin to Alias Table
+
+    #{MarkdownTableGenerator.generate(alias_tuples, true)}
```
Why is `alias_to_builtin` true, for "Builtin to Alias Table"?
`MarkdownTableGenerator.generate` seems very confusing and unnecessarily complicated.

keithrbennett replied:
Regarding the file location, you're right, I know you mentioned this a while back and I thought I fixed it, but I guess I didn't. Regarding the 'alias_to_builtin' flag, I inadvertently reversed its sense. If that had been a keyword argument it would have been more obvious. Would you like me to change it to a keyword argument, and change the name to 'builtin_to_alias'?

I agree with you that it's a bit complex, due to my wanting not to duplicate much of the code in separate methods for the two directions. I can look into simplifying this if you like. If you want to remove this script from the code base entirely and just rely on manual updating of the readme, that is fine with me too.

This is really embarrassing! I apologize for the errors.

paddor commented:
What about `$"` and `$:`? I did not see them.

keithrbennett commented:
@paddor This issue is about documenting the aliases in tables in the readme. The aliases you mention are not included in this gem, so they are not included in the documentation tables.

If you mean to suggest that they be added to the gem I suggest posting a new issue if it hasn't been addressed in a previous issue.
